File: 2016/day02.cpp
// https://adventofcode.com/2016/day/2

#include "../common/common.h"
#include "../common/rust.h"

#ifndef AOC_IMPORT_STD
#include <array>
#include <print>
#endif

using Input = Vec<String>;

auto parse(String const& filename) -> Input {
  return aoc::read_lines(filename);
}

fn to_diff(char direction) -> point {
  switch (direction) {
    case 'U':
      return aoc::get_diff<int>(aoc::north);
    case 'D':
      return aoc::get_diff<int>(aoc::south);
    case 'L':
      return aoc::get_diff<int>(aoc::west);
    case 'R':
      return aoc::get_diff<int>(aoc::east);
    default:
      AOC_UNREACHABLE("Invalid direction");
  }
}

template <bool Complicated>
fn solve_case(Input const& input) -> String {
  constexpr let keypad = [] {
    if constexpr (!Complicated) {
      return aoc::fixed_char_grid<3, 3>("123"
                                        "456"
                                        "789");
    } else {
      return aoc::fixed_char_grid<5, 5>("  1  "
                                        " 234 "
                                        "56789"
                                        " ABC "
                                        "  D  ");
    }
  }();

  auto pos = aoc::as_consteval(
      keypad.position(aoc::ranges::position(keypad, '5').value()));
  auto code = String{};

  for (let& line : input) {
    for (let direction : line) {
      let new_pos = pos + to_diff(direction);
      if (keypad.in_bounds(new_pos.y, new_pos.x) &&
          (!Complicated || (keypad.at(new_pos.y, new_pos.x) != ' '))) {
        pos = new_pos;
      }
    }
    code.push_back(keypad.at(pos.y, pos.x));
  }

  return code;
}

int main() {
  std::println("Part 1");
  let example = parse("day02.example");
  AOC_EXPECT_RESULT("1985", solve_case<false>(example));
  let input = parse("day02.input");
  AOC_EXPECT_RESULT("73597", solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT("5DB3", solve_case<true>(example));
  AOC_EXPECT_RESULT("A47DA", solve_case<true>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2016/src/day02.rs
use aoc::grid::{ConfigInput, Grid, Ipos};

type Input = Vec<String>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
}

fn solve_case<const COMPLICATED: bool>(input: &Input) -> String {
    let keypad_str = if !COMPLICATED {
        vec![
            "123", //
            "456", //
            "789", //
        ]
    } else {
        vec![
            "  1  ", //
            " 234 ", //
            "56789", //
            " ABC ", //
            "  D  ", //
        ]
    };

    let (keypad, config) = Grid::from_lines_config(
        &keypad_str,
        ConfigInput {
            start_char: Some('5'),
            ..Default::default()
        },
    );
    let pos = config.start_pos.unwrap();
    let mut pos = Ipos::new(pos.x as isize, pos.y as isize);
    let mut code = String::new();

    for line in input {
        for dir in line.chars() {
            let diff = match dir {
                'U' => Ipos::new(0, -1),
                'D' => Ipos::new(0, 1),
                'L' => Ipos::new(-1, 0),
                'R' => Ipos::new(1, 0),
                _ => unreachable!("Invalid direction"),
            };
            let new_pos = pos + diff;
            if keypad.in_bounds_signed(new_pos.y, new_pos.x)
                && (!COMPLICATED || (*keypad.get(new_pos.y as usize, new_pos.x as usize) != ' '))
            {
                pos = new_pos;
            }
        }
        code.push(*keypad.get(pos.y as usize, pos.x as usize))
    }

    code
}

fn main() {
    println!("Part 1");
    let example = parse("day02.example");
    aoc::expect_result!("1985", solve_case::<false>(&example));
    let input = parse("day02.input");
    aoc::expect_result!("73597", solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!("5DB3", solve_case::<true>(&example));
    aoc::expect_result!("A47DA", solve_case::<true>(&input));
}
